File: leetCode/fukun/001-100/lc_098.js
var isValidBST = function (root) {
    const helper = function (node, tl, tr) {
        if (!node) return true;
        return (!node.left || (node.left.val > tl && node.left.val < node.val))
        && (!node.right || (node.right.val < tr && node.right.val > node.val))
            ? helper(node.left, tl, node.val) && helper(node.right, node.val, tr) : false;
    };
    return helper(root, -Infinity, Infinity);
};

File: leetCode/lwc/0001-0100/lc_098.java
package solution2;

import java.util.ArrayList;
import java.util.List;

/**
 * @author lwc
 * @date 2019/10/26 23:29
 */
public class U98 {
    public boolean isValidBST(TreeNode root) {
        if(root == null) return true;
        return recursive(root,new ArrayList<>()) == 1;
    }

    public int recursive(TreeNode root, List<Integer> v) {
        int flag = 1;
        if(root.left != null) flag &= recursive(root.left, v);
        if (!v.isEmpty() && root.val <= v.get(v.size()-1)) return 0;
        v.add(root.val);
        if(root.right != null) flag &= recursive(root.right, v);

        return flag;
    }

    public static void main(String[] args) {
        new U98().isValidBST(new TreeNode(0));
    }
}
